### DISCUSSION TITLE
Resources on Ethereum Internals

### DISCUSSION THREAD

**Discussion by maxzaver** [near/nearcore#3733] [Resources]
How Ethereum serializes logs: https://medium.com/linum-labs/everything-you-ever-wanted-to-know-about-events-and-logs-on-ethereum-fec84ea7d0a5

**Comment by ailisp** [comment]
RLP spec:
https://eth.wiki/fundamentals/rlp

RLP online encoder/decoder:
https://toolkit.abdk.consulting/ethereum#rlp
https://codechain-io.github.io/rlp-debugger/

RLP Rust implementation (pass identity closure as f to decode, see test):
https://github.com/paritytech/parity-common/blob/master/rlp/src/rlpin.rs#L396

ABI spec:
https://docs.soliditylang.org/en/v0.7.5/abi-spec.html

EIP-712 spec:
https://eips.ethereum.org/EIPS/eip-712

EIP-712 Rust implementation (algorithm can be referenced in our implementation, however note the msg is in json string, ours are separate rust variables and function args in ABI-encoded u8 array; to by skip json, function and types in this crate is not `pub`; so using their algo, adapt with our data structure would be best)
https://github.com/openethereum/openethereum/blob/main/util/EIP-712/src/encode.rs
